src/routes/subscriptions.rs

```diff
@@ -32,59 +32,126 @@ impl TryFrom<FormData> for NewSubscriber {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            SubscribeError::StoreToken(_) => write!(
+                f,
+                "Failed to store the confirmation token for a new subscriber."
+            ),
+            SubscribeError::SendEmail(_) => write!(f, "Failed to send a confirmation email."),
+            SubscribeError::Validation(e) => write!(f, "{}", e),
+            SubscribeError::InsertSubscriber(_) => {
+                write!(f, "Failed to insert new subscriber in the database.")
+            }
+            SubscribeError::Pool(_) => {
+                write!(f, "Failed to acquire a Postgres connection from the pool.")
+            }
+            SubscribeError::TransactionCommit(_) => write!(
+                f,
+                "Failed to commit SQL transaction to store a new subscriber"
+            ),
+        }
+    }
+}
+
+impl IntoResponse for SubscribeError {
+    fn into_response(self) -> axum::response::Response {
+        match self {
+            SubscribeError::Validation(_) => http::StatusCode::BAD_REQUEST.into_response(),
+            SubscribeError::InsertSubscriber(_)
+            | SubscribeError::Pool(_)
+            | SubscribeError::TransactionCommit(_)
+            | SubscribeError::SendEmail(_)
+            | SubscribeError::StoreToken(_) => {
+                http::StatusCode::INTERNAL_SERVER_ERROR.into_response()
+            }
+        }
+    }
+}
+
+impl From<reqwest::Error> for SubscribeError {
+    fn from(e: reqwest::Error) -> Self {
+        Self::SendEmail(e)
+    }
+}
+
+impl From<StoreTokenError> for SubscribeError {
+    fn from(e: StoreTokenError) -> Self {
+        Self::StoreToken(e)
+    }
+}
+
+impl From<String> for SubscribeError {
+    fn from(e: String) -> Self {
+        Self::Validation(e)
+    }
+}
+
 #[tracing::instrument(
     name = "Adding a new subscriber",
     skip(form, app_state),
     fields(
         subscriber_email = %form.email,
         subscriber_name = %form.name,
-    )
+    ),
+    err(Debug),
 )]
 pub async fn subscribe(
     State(app_state): State<Arc<AppState>>,
     Form(form): Form<FormData>,
-) -> impl IntoResponse {
-    let new_subscriber = match form.try_into() {
-        Ok(new_subscriber) => new_subscriber,
-        Err(_) => return StatusCode::BAD_REQUEST,
-    };
-
-    let mut transaction = match app_state.pool.begin().await {
-        Ok(transaction) => transaction,
-        Err(_) => return StatusCode::INTERNAL_SERVER_ERROR,
-    };
-
-    let subscriber_id = match insert_subscriber(&mut transaction, &new_subscriber).await {
-        Ok(subscriber_id) => subscriber_id,
-        Err(_) => return StatusCode::INTERNAL_SERVER_ERROR,
-    };
-    let subscription_token = generate_subscription_token();
+) -> Result<impl IntoResponse, SubscribeError> {
+    let new_subscriber = form.try_into()?;
+    let mut transaction = app_state.pool.begin().await.map_err(SubscribeError::Pool)?;
```

Comment:
Since this line, `insert_subscriber`, and committing the transaction all return the possible `sqlx::Error`, we have to explicitly map these